lib/src/code_field/code_controller.dart

```diff
@@ -380,6 +384,121 @@ class CodeController extends TextEditingController {
+  void commentOrUncommentSelection() {
+    final commentTypes = SingleLineComments.byMode[language] ?? [];
+
+    for (final commentType in commentTypes) {
+      final regExpForCommentedLine =
+          RegExps.getSingleLineCommentRegExp(commentType, multiline: true);
+
+      // line is either a comment or empty
+      final allSelectedLinesAreCommented = checkSelectedLinesForCondition(
+        (line) =>
+            regExpForCommentedLine.hasMatch(line) ||
+            RegExps.emptyLine.hasMatch(line),
+      );
+
+      if (allSelectedLinesAreCommented) {
+        _uncommentSelectedLines(commentType);
+        return;
+      } else {
+        _commentSelectedLines(commentType);
+        return;
+      }
+    }
+  }
+
+  /// Utility method to divide [commentOrUncommentSelection] method into parts
+  /// [commentSymbols] - `//` or `#` indicates what to use to comment
+  void _commentSelectedLines(String commentSymbols) {
+    int? startOfCommentSymbols;
+
+    modifySelectedLines((line) {
+      // if line is empty do not comment it
+      if (RegExps.emptyLine.hasMatch(line)) {
+        return line;
+      }
+
+      startOfCommentSymbols ??=
+          RegExps.whiteSpacesAfterStartOfLine.firstMatch(line)?.end;
+
+      if (startOfCommentSymbols == null) {
+        return '$commentSymbols $line';
+      }
+
+      return line.replaceRange(
+        startOfCommentSymbols!,
+        startOfCommentSymbols,
+        '$commentSymbols ',
+      );
+    });
+  }
+
+  /// Utility method to divide [commentOrUncommentSelection] method into parts
+  /// [commentSymbols] - `//` or `#` indicates what to use to uncomment
+  void _uncommentSelectedLines(String commentSymbols) {
+    modifySelectedLines((line) {
+      // if line is empty skip it
+      if (RegExps.emptyLine.hasMatch(line)) {
+        return line;
+      }
+
+      // replace first `// ` or `# ` or `//` or `#` with empty string
+      return line.replaceFirst(
+        RegExps.getCommentPlusWhitespaceRegExp(commentSymbols),
+        '',
+      );
+    });
+  }
+
+  /// Returns true if all selected lines meet condition in the callback.
+  /// Returns false otherwise.
+  bool checkSelectedLinesForCondition(bool Function(String line) callback) {
+    if (selection.start == -1 || selection.end == -1) {
+      return false;
+    }
+
+    final selectionStart = _code.hiddenRanges.recoverPosition(
+      selection.start,
+      placeHiddenRanges: TextAffinity.downstream,
+    );
+    final selectionEnd = _code.hiddenRanges.recoverPosition(
+      // to avoid including the next line if `\n` is selected
+      selection.isCollapsed ? selection.end : selection.end - 1,
+      placeHiddenRanges: TextAffinity.downstream,
+    );
+
+    // index of the first line to modify
+    final firstLineIndex =
+        _code.lines.characterIndexToLineIndex(selectionStart);
+    // index of the last line to modify
+    final lastLineIndex = _code.lines.characterIndexToLineIndex(selectionEnd);
```

Comment:
Extract to `TextRange getSelectedLineRange()`, reuse it in `modifySelectedLines`.

`TextRange.start` should be the first selected line. `TextRange.end` should be the last selected line + 1.